### ISSUE TITLE
Rule to enforce Js "classes" declaration

### ISSUE THREAD

**Issue by htulipe** [eslint/eslint#2630]
This more a question than an issue. I would like to prevent people from doing this:

	var MyClass = function() {
		this.sayHello = function() {
			console.log("hello");
		};
	};

Instead, they should do this:

	var MyClass = function() {};
	MyClass.prototype.sayHello = function() {
		console.log("hello");
	};

Am I correct by saying there is no built-in rule for enforcing such a behavior in eslint? I never wrote custom rules, how difficult such a rule would be to create?

**Comment by ilyavolodin** [comment]
You are correct, we do not have a rule like that. I also think this rule is too specific for us to add to the core, but it wouldn't be hard to build it yourself as a plugin. You can take a look at the code for some of our existing rules, and read through documentation here: http://eslint.org/docs/developer-guide/working-with-rules. I'm going to close this issue, but if you have trouble creating this rule, feel free to join our [gitter](https://gitter.im/eslint/eslint) room and ask for help there.

**Comment by xjamundx** [comment]
@htulipe it would probably look something like this:

```js
module.exports = function(content) {
    "AssignmentExpression": function(node) {
        if (node.left.type === "MemberExpression" &&
            node.left.object.type === "ThisExpression" &&
            node.right.type === "FunctionExpression") {
                context.report(node);
        }
    }
}
```

You build rules around the abstract syntax tree. Paste the code you want to avoid here to get a sense of what sorts of nodes you need to check for:
http://esprima.org/demo/parse.html

For example your code example you're against looks like this:
```js
{
    "type": "Program",
    "body": [
        {
            "type": "VariableDeclaration",
            "declarations": [
                {
                    "type": "VariableDeclarator",
                    "id": {
                        "type": "Identifier",
                        "name": "MyClass"
                    },
                    "init": {
                        "type": "FunctionExpression",
                        "id": null,
                        "params": [],
                        "defaults": [],
                        "body": {
                            "type": "BlockStatement",
                            "body": [
                                {
                                    "type": "ExpressionStatement",
                                    "expression": {
                                        "type": "AssignmentExpression",
                                        "operator": "=",
                                        "left": {
                                            "type": "MemberExpression",
                                            "computed": false,
                                            "object": {
                                                "type": "ThisExpression"
                                            },
                                            "property": {
                                                "type": "Identifier",
                                                "name": "sayHello"
                                            }
                                        },
                                        "right": {
                                            "type": "FunctionExpression",
                                            "id": null,
                                            "params": [],
                                            "defaults": [],
                                            "body": {
                                                "type": "BlockStatement",
                                                "body": [
                                                    {
                                                        "type": "ExpressionStatement",
                                                        "expression": {
                                                            "type": "CallExpression",
                                                            "callee": {
                                                                "type": "MemberExpression",
                                                                "computed": false,
                                                                "object": {
                                                                    "type": "Identifier",
                                                                    "name": "console"
                                                                },
                                                                "property": {
                                                                    "type": "Identifier",
                                                                    "name": "log"
                                                                }
                                                            },
                                                            "arguments": [
                                                                {
                                                                    "type": "Literal",
                                                                    "value": "hello",
                                                                    "raw": "\"hello\""
                                                                }
                                                            ]
                                                        }
                                                    }
                                                ]
                                            },
                                            "generator": false,
                                            "expression": false
                                        }
                                    }
                                }
                            ]
                        },
                        "generator": false,
                        "expression": false
                    }
                }
            ],
            "kind": "var"
        }
    ]
}
```

**Comment by htulipe** [comment]
@xjamundx Thanks for the code sample and link. I thought it would take a longer piece of code.
@ilyavolodin Thanks for your advice. I must say I disagree when you say it is a too specific rule. The code I am trying to avoid is really bad practice because a new instance of every function will be created in memory each time you instanciate the "class". Whereas, with the prototype, only one instance will be created and shared among all instances. Some code to clarify :

    var item1 = new MyClassUsingPrototype();
    var item2 = new MyClassUsingPrototype();
    console.log(item1.sayHello === item2.sayHello); // True

But:

    var item1 = new MyClassUsingThis();
    var item2 = new MyClassUsingThis();
    console.log(item1.sayHello === item2.sayHello); // False

In a case where you instanciate lot of objects, it can lead to severe performance issues.

**Comment by ilyavolodin** [comment]
@htulipe I understand that. However, the reason I said it's too generic, is because there are a lot of people who don't want prototypical behavior. So they would assign properties to an object directly, instead of doing it on the prototype. For example, whenever I create classes that way, I always create `initialize` method on the object, not on the prototype.
We don't have any way to determine what the user wanted in each case, that's why the rule is too generic. It's a good rule for plugin, if you are willing to follow it all the time, but I don't think it's good fit for the core.

**Comment by htulipe** [comment]
I get your point and will create a plugin. Thanks again for your reactivity and for the eslint project, it is great.

**Comment by htulipe** [comment]
For the record, I pushed a first draft here: https://github.com/htulipe/eslint-plugin-no-this-func
